--- /arch/arm/boot/dts/bcm2710-rpi-3-b.dts	2019-07-30 08:37:58.000000000 -0500
+++ /arch/arm/boot/dts/bcm2710-rpi-3-b.dts	2019-08-28 16:24:02.950104259 -0500
@@ -25,10 +25,20 @@
 	};
 
 	spi0_cs_pins: spi0_cs_pins {
-		brcm,pins = <8 7>;
+		brcm,pins = <8>;
 		brcm,function = <1>; /* output */
 	};
 
+	spi1_pins: spi1_pins {
+		brcm,pins = <19 20 21>;
+		brcm,function = <3>; /* alt4 */
+	};
+
+	spi1_cs_pins: spi1_cs_pins {
+		brcm,pins = <18>;
+		brcm,function = <1>; /* output */
+	};	
+
 	i2c0_pins: i2c0 {
 		brcm,pins = <0 1>;
 		brcm,function = <4>;
@@ -55,17 +65,22 @@
 		brcm,function = <4>; /* alt0:GPCLK2 */
 		brcm,pull = <0>;
 	};
+	
+	fake_bt_cts: fake_bt_cts {
+		brcm,pins = <31>;
+		brcm,function = <1>; /* output */
+	};
 
 	uart0_pins: uart0_pins {
-		brcm,pins = <32 33>;
-		brcm,function = <7>; /* alt3=UART0 */
+		brcm,pins = <14 15>;
+		brcm,function = <4>; /* alt0=UART0 */
 		brcm,pull = <0 2>;
 	};
 
 	uart1_pins: uart1_pins {
-		brcm,pins;
-		brcm,function;
-		brcm,pull;
+		brcm,pins = <32 33>;
+		brcm,function = <2>; /* alt5=UART1 */
+		brcm,pull = <0 2>;
 	};
 
 	audio_pins: audio_pins {
@@ -103,38 +118,76 @@
 
 &uart0 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&uart0_pins &bt_pins>;
+	pinctrl-0 = <&uart0_pins>;
 	status = "okay";
 };
 
 &uart1 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&uart1_pins>;
+	pinctrl-0 = <&uart1_pins &bt_pins &fake_bt_cts>;
 	status = "okay";
 };
 
 &spi0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
-	cs-gpios = <&gpio 8 1>, <&gpio 7 1>;
+	cs-gpios = <&gpio 8 1>;
+	status = "okay";
 
-	spidev0: spidev@0{
-		compatible = "spidev";
-		reg = <0>;	/* CE0 */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		spi-max-frequency = <125000000>;
-	};
-
-	spidev1: spidev@1{
-		compatible = "spidev";
-		reg = <1>;	/* CE1 */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		spi-max-frequency = <125000000>;
+	sc16is762: sc16is762@0 {
+		compatible = "nxp,sc16is762";
+		reg = <0>; /* CE0 */
+		clocks = <&sc16is762_clk>;
+		interrupt-parent = <&gpio>;
+		interrupts = <23 2>; /* IRQ_TYPE_EDGE_FALLING */
+		#gpio-controller;
+		#gpio-cells = <2>;
+		spi-max-frequency = <15000000>;
+
+		sc16is762_clk: sc16is762_clk {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <8000000>;
+		};
 	};
 };
 
+&spi1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi1_pins &spi1_cs_pins>;
+	cs-gpios = <&gpio 18 1>;
+	status = "okay";
+
+	sc16is762: sc16is762@0 {
+		compatible = "nxp,sc16is762";
+		reg = <0>; /* CE0 */
+		clocks = <&sc16is762_clk>;
+		interrupt-parent = <&gpio>;
+		interrupts = <24 2>; /* IRQ_TYPE_EDGE_FALLING */
+		#gpio-controller;
+		#gpio-cells = <2>;
+		spi-max-frequency = <15000000>;
+
+		sc16is762_clk: sc16is762_clk {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <8000000>;
+		};
+	};
+};
+
+&aux {
+	status = "okay";
+};
+
+&watchdog {
+	status = "okay";
+};		
+
 &i2c0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&i2c0_pins>;
@@ -145,6 +198,13 @@
 	pinctrl-names = "default";
 	pinctrl-0 = <&i2c1_pins>;
 	clock-frequency = <100000>;
+	status = "okay";
+	
+	pcf8523: pcf8523@68 {
+		compatible = "nxp,pcf8523";
+		reg = <0x68>;
+		status = "okay";
+	};		
 };
 
 &i2c2 {
@@ -165,7 +225,7 @@
 
 	pwr_led: pwr {
 		label = "led1";
-		linux,default-trigger = "input";
+		linux,default-trigger = "heartbeat";
 		gpios = <&expgpio 7 0>;
 	};
 };
